onnx/utils.py

```diff
@@ -21,3 +24,154 @@ def polish_model(model):  # type: (ModelProto) -> ModelProto
     model = onnx.optimizer.optimize(model)
     onnx.checker.check_model(model)
     return model
+
+
+class Extractor:
+    def __init__(self, model):  # type: (ModelProto) -> None
+        self.model = onnx.shape_inference.infer_shapes(model)
+        self.graph = self.model.graph
+        self.wmap = self._build_name2obj_dict(self.graph.initializer)
+        self.vimap = self._build_name2obj_dict(self.graph.value_info)
+
+    @staticmethod
+    def _build_name2obj_dict(objs):  # type: ignore
+        return {obj.name: obj for obj in objs}
+
+    def _collect_new_io_core(self, original_io, io_names_to_extract):  # type: ignore
+        original_io_map = self._build_name2obj_dict(original_io)
+        original_io_names = set(original_io_map.keys())
+        s_io_names_to_extract = set(io_names_to_extract)
+        io_names_to_keep = s_io_names_to_extract & original_io_names
+        new_io_names_to_add = s_io_names_to_extract - original_io_names
+
+        new_io_tensors = []
+        for name in io_names_to_keep:
+            new_io_tensors.append(original_io_map[name])
+        for name in new_io_names_to_add:
+            # activation become input or output
+            new_io_tensors.append(self.vimap[name])
+
+        # adjust sequence
+        new_io_tensors_map = self._build_name2obj_dict(new_io_tensors)
+        return [new_io_tensors_map[name] for name in io_names_to_extract]
+
+    def _collect_new_inputs(self, names):  # type: (List[Text]) -> List[ValueInfoProto]
+        return self._collect_new_io_core(self.graph.input, names)  # type: ignore
+
+    def _collect_new_outputs(self, names):  # type: (List[Text]) -> List[ValueInfoProto]
+        return self._collect_new_io_core(self.graph.output, names)  # type: ignore
+
+    def _dfs_search_reachable_nodes(
+            self,
+            node_output_name,  # type: Text
+            graph_input_names,  # type: List[Text]
+            reachable_nodes,  # type: List[NodeProto]
+    ):  # type: (...) -> None
+        if node_output_name in graph_input_names:
+            return
+        for node in self.graph.node:
+            if node in reachable_nodes:
+                continue
+            if node_output_name not in node.output:
+                continue
+            reachable_nodes.append(node)
+            for name in node.input:
+                self._dfs_search_reachable_nodes(name, graph_input_names, reachable_nodes)
+
+    def _collect_reachable_nodes(
+            self,
+            input_names,  # type: List[Text]
+            output_names,  # type: List[Text]
+    ):  # type: (...) -> List[NodeProto]
+        reachable_nodes = list()  # type: ignore
+        for name in output_names:
+            self._dfs_search_reachable_nodes(name, input_names, reachable_nodes)
+        # needs to be topology sorted.
+        nodes = [n for n in self.graph.node if n in reachable_nodes]
+        return nodes
+
+    def _collect_reachable_tensors(
```

Comment:
To optimize I think this can be done along with _collect_reachable_nodes. This is nice to have.

Reply:
Yes, we can do that. IMO, _two passes_ is sort clearer to clarify the logic. And, as `_collect_reachable_tensors` doesn't walk the graph but iterate nodes, I guess it won't be significant :) Anyway, I can rewrite it if we really want that.

Reply:
As I mentioned this is not a must have but a nice to have.

Reply:
Let's see if we can improve this in the future :)